



cd /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/
OUTDIR=/groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/
ml samtools


samtools merge "$OUTDIR"/finalBamFile.bam  "$OUTDIR"/combinedFile_CGGTTA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TTAACT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_ATGAAC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CCTAAG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_AATCCG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GGCTGC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TACCTT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TCTTAA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GTCAGG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_ATACTG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TATGTC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GAGTCC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GGAGGT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CACACT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CCGCAA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TTTATG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_AACGCC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CAAGCA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam
samtools mpileup -B -A -l /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/allStagesCombined/countingWindows_transcriptionalOutput_priMirna.bed -f /groups/ameres/bioinformatics/references/danio_rerio/dr10/danRer10.fa  "$OUTDIR"/finalBamFile.bam >  "$OUTDIR"/finalBam_R1_pileup.mpileup
cd /groups/ameres/bioinformatics/tools/slamdunk_new-2/slamdunk/slamdunk/contrib/
java -jar VarScan.v2.4.1.jar  mpileup2snp /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/finalBam_R1_pileup.mpileup  --min-coverage 10 --min-var-freq 0.20 --strand-filter 0 --variants 1 > /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/varSample_R1.vcf


### replicate 2

samtools merge  "$OUTDIR"/finalBamFile_R2.bam  "$OUTDIR"/combinedFile_GCTCGA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GCGAAT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TGGATT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_ACCTAC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CGAAGG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_AGATAG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TTGGTA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GTTACC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CGCAAC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TGGCGA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_ACCGTG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CAACAG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_GATTGT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_CTCTCG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TGACAC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_AAGACA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_ACAGAT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam  "$OUTDIR"/combinedFile_TAGGCT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam
samtools mpileup -B -A -l /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/allStagesCombined/countingWindows_transcriptionalOutput_priMirna.bed -f /groups/ameres/bioinformatics/references/danio_rerio/dr10/danRer10.fa  "$OUTDIR"/finalBamFile_R2.bam >  "$OUTDIR"/finalBam_R2_pileup.mpileup
cd /groups/ameres/bioinformatics/tools/slamdunk_new-2/slamdunk/slamdunk/contrib/
java -jar VarScan.v2.4.1.jar  mpileup2snp /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/finalBam_R2_pileup.mpileup  --min-coverage 10 --min-var-freq 0.20 --strand-filter 0 --variants 1 > /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/varSample_R2.vcf

##replica 3
samtools merge "$OUTDIR"/finalBamFile_R3.bam "$OUTDIR"/combinedFile_CTCCAT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_GCATGG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_AATAGC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_GTGCCA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_TCGAGG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_CACTAA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_GGTATA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_CGCCTG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_AATGAA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_ACAACG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_ATATCC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_AGTACT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_ATAAGA.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_GGTGAG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_TTCCGC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_GAAGTG.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_CAATGC.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam "$OUTDIR"/combinedFile_ACGTCT.fastq.gz_adapterTrimmed_slamdunk_mapped_filtered.bam
samtools mpileup -B -A -l /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/allStagesCombined/countingWindows_transcriptionalOutput_priMirna.bed -f /groups/ameres/bioinformatics/references/danio_rerio/dr10/danRer10.fa  "$OUTDIR"/finalBamFile_R3.bam >  "$OUTDIR"/finalBam_R3_pileup.mpileup
cd /groups/ameres/bioinformatics/tools/slamdunk_new-2/slamdunk/slamdunk/contrib/
java -jar VarScan.v2.4.1.jar  mpileup2snp /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/finalBam_R3_pileup.mpileup  --min-coverage 10 --min-var-freq 0.20 --strand-filter 0 --variants 1 > /groups/ameres/Pooja/Projects/zebrafishAnnotation/sequencingRun_december2017/analysis/annotation/stageSpecific/outputs/slamDunk_combinedStageSpecific/filter/varSample_R3.vcf




